crates/polars-io/src/cloud/polars_object_store.rs

```diff
@@ -318,28 +319,30 @@ fn merge_ranges(ranges: &[Range<usize>]) -> impl Iterator<Item = (Range<usize>,
                     (r.abs_diff(l), r < l)
                 };
 
-                #[rustfmt::skip]
-                let should_merge =
-                    is_overlapping // Always merge if overlapping
-                    || (
-                        (
-                            // Either one range is extremely small compared to the other, with a limit of 8MiB..
-                            range.len().min(current_merged_range.len())
-                                < (range.len().max(current_merged_range.len()) / 128).min(8 * 1024 * 1024)
-                            // ..or the new size is closer to the chunk_size
-                            || new_merged.len().abs_diff(chunk_size) < current_merged_range.len().abs_diff(chunk_size)
-                        )
-                        && (
-                            // Either the gap is less than 1MiB..
-                            distance <= 1024 * 1024
-                            || (
-                                // ..or, the gap is less than 12.5% of the largest between `current_n_bytes`
-                                // and the new `range`, capped at 8MiB.
-                                distance <= current_n_bytes.max(range.len()) / 8
-                                && distance <= 8 * 1024 * 1024
-                            )
-                        )
-                    );
+                // * Always merge if overlapping
+                // * Otherwise, merge only if the distance between the ranges is within tolerance, and either:
+                //   * The size of the new merged range is closer to the chunk size,
+                //   * Or, one side is relatively small compared to the other. The threshold is weighted by the number
+                //     of already-merged items to ensure a logarithmic worst-case growth rate.
+                let should_merge = is_overlapping || {
+                    #[allow(non_snake_case)]
+                    let distance_within_1MiB_or_12perc_max_8MiB = distance
+                        <= (current_n_bytes.max(range.len()) / 16)
+                            .clamp(1024 * 1024, 8 * 1024 * 1024);
+
+                    let new_merged_is_closer_to_chunk_size = new_merged.len().abs_diff(chunk_size)
+                        <= current_merged_range.len().abs_diff(chunk_size);
+
+                    let one_side_is_relatively_small = {
+                        let fraction = range.len().max(current_merged_range.len())
+                            / current_n_merged.saturating_mul(16);
+
+                        range.len().min(current_merged_range.len()) <= fraction
+                    };
+
+                    distance_within_1MiB_or_12perc_max_8MiB
```

Comment:
the one massive statement broke formatting for the whole file 🥲, so I've just split it to descriptively named variables instead